### PR TITLE
refactor: casl-factory - instruments

### PR DESCRIPTION
**Pull request by HayenNico**
## Description
Subsection of PR #2748 for instruments. PR depends on #2759 to be merged first.

This transfers the instrumentEndpointAccess function in CaslAbilityFactory to the refactored format, and adjusts the affected controller (datasets-access) to accommodate the change. The instrument-specific auth code is extracted into a separate module.

## Changes:

* Replace CaslAbilityFactory.instrumentEndpointAccess with CaslAbilityFactory.instrumentAccess
* Code for CaslAbilityFactory.instrumentAccess is factored out into new module InstrumentAbility
* Adjust endpoint logic in one function in datasets-access controller

## Tests included

- [x] Included for each change/fix?
- [x] Passing? 

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

## Summary by Sourcery

Refactor instrument access control to use a dedicated InstrumentAbility and update consumers accordingly.

Enhancements:
- Extract instrument authorization rules into a new InstrumentAbility service and wire it into the Casl module and factory.
- Simplify instrument access logic to grant read to all users while gating create, update, and delete by admin and delete groups.
- Update dataset access service to use the new instrument access entry point and align Action enum layout for instrument-related actions.

### CODE REVIEW

File: src/casl/abilities/instruments.ability.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    if (user.currentGroups.some((g) => this.accessGroups?.admin.includes(g))) {
+      /**
+       * User belonging to ADMIN_GROUPS
+       */
+      can(Action.InstrumentCreate, Instrument);
+      can(Action.InstrumentUpdate, Instrument);
+    }
+
+    if (user.currentGroups.some((g) => this.accessGroups?.delete.includes(g))) {
```
[the remark this answers is not in the document]
Comment (HayenNico) [reply src/casl/abilities/instruments.ability.ts]: Will be resolved at a later stage by factoring the accessGroups membership checks out into a common helper

**Comment by nitrosx** [comment]
# Code Review Report: refactor-casl-factory-instruments Branch

---
## Overview

### 1. Logic Correctness
The refactoring maintains identical authorization logic for instruments. The permissions are correctly extracted from `CaslAbilityFactory.instrumentEndpointAccess` to `InstrumentAbility.buildAbility` with no functional changes. The method signature changes from `(user: JWTUser)` to `(user: JWTUser | null)` to properly support unauthenticated users, which is an improvement.

### 2. Edge Cases Handled
- Unauthenticated users: Can only read instruments
- Authenticated users: Can only read instruments (no write access by default)
- ADMIN_GROUPS members: Can read, create, and update instruments
- DELETE_GROUPS members: Can read and delete instruments
- The logic correctly handles the case where `accessGroups` configuration is missing (falls back to empty object)

### 3. What the Code Does
The changes refactor the CASL (Authorization) module by:
- Extracting instrument-specific authorization logic into a dedicated `InstrumentAbility` class
- Moving instrument actions in `Action` enum to a more consistent location (grouped with other resource types)
- Updating all references to use the new `instrumentAccess` method instead of `instrumentEndpointAccess`
- Adding comprehensive documentation for the instruments authorization model

### 4. Assessment of Changes
The changes make sense as part of a broader refactoring effort to modularize the CASL authorization system. The pattern follows the existing `DatasetAbility` extraction, creating consistency across the codebase.

### 5. Unreachable Code
No unreachable code identified. The removed `instrumentEndpointAccess` method in `CaslAbilityFactory` was fully replaced by the new `InstrumentAbility` class and `instrumentAccess` wrapper method.

---
## Code Changes

### Summary of Changes

| File | Change Type | Description |
|------|-------------|-------------|
| `src/casl/abilities/instruments.ability.ts` | New | Created dedicated ability class for instruments |
| `src/casl/action.enum.ts` | Modified | Moved instrument actions from bottom to top with other resource types |
| `src/casl/casl-ability.factory.ts` | Modified | Removed inline `instrumentEndpointAccess`, added `InstrumentAbility` dependency, added `instrumentAccess` wrapper |
| `src/casl/casl.module.ts` | Modified | Added `InstrumentAbility` to providers |
| `src/casl/casl-ability.factory.spec.ts` | Modified | Updated tests to inject `InstrumentAbility` |
| `src/datasets/datasets-access.service.ts` | Modified | Changed method call from `instrumentEndpointAccess` to `instrumentAccess` |
| `docs/developer-guide/authorization/instruments.md` | New | Added comprehensive authorization documentation |

### Assessment
**Necessary**: Yes. The refactoring improves code organization, maintainability, and follows established patterns in the codebase (similar to `DatasetAbility`). The extraction reduces the size of the monolithic `CaslAbilityFactory` class and makes instrument authorization logic more testable and maintainable.

### Improvement Needed
1. **Test Coverage**: The new `InstrumentAbility` class lacks dedicated unit tests. While the factory tests compile, there are no tests specifically validating the authorization rules in `InstrumentAbility.buildAbility`.
2. **Documentation Consistency**: The new documentation file is comprehensive but should follow any existing documentation style guidelines.
3. **Action Enum Ordering**: While moving instrument actions to the top is more logical, consider alphabetical ordering or grouping by domain for consistency across the enum.

### Verdict
**Positive**. The refactoring is well-executed, maintains backward compatibility, and follows established patterns. The extraction of concerns improves code maintainability without changing functionality.

Commits responsible:
- `31e2952f`: Initial refactoring with rebase fixes
- `91b9b2fd`: Fixed unit test initialization

---
## Security Review

### 1. Potential Injection Vulnerabilities
**No injection vulnerabilities identified**. The authorization logic uses CASL's type-safe ability builder pattern. All user input is validated through the JWT authentication middleware before reaching these authorization checks. The `user.currentGroups` values are arrays of strings from authenticated tokens, not raw user input.

### 2. Sensitive User Data Exposure
**No sensitive data exposure**. The authorization logic only checks group memberships and permission levels. It does not expose any user data, instrument data, or other sensitive information through the authorization mechanisms.

### 3. Insecure API Usage
**No insecure API usage identified**. The CASL library is used correctly according to its documentation. The `MongoAbility` type provides type safety for MongoDB queries.

### 4. Authentication Bypass
**No authentication bypass possible**. The logic correctly:
- Denies all write operations for unauthenticated users
- Requires specific group memberships (ADMIN_GROUPS or DELETE_GROUPS) for write operations
- Uses the `currentGroups` from the JWT token which is validated by the authentication middleware
- Maintains the same permission checks as the original implementation

---
## Test Coverage

### Current State
- Factory-level tests in `casl-ability.factory.spec.ts` verify that the factory can be instantiated with `InstrumentAbility`
- No dedicated tests for `InstrumentAbility.buildAbility` authorization rules
- No tests verifying the specific permission matrix (unauthenticated read-only, ADMIN_GROUPS create/update, DELETE_GROUPS delete)

### Adequacy Assessment
**Inadequate**. While the refactoring maintains the same logic, there are no tests to verify:
1. Unauthenticated users can only read instruments
2. Authenticated users without special groups can only read instruments
3. ADMIN_GROUPS members can create and update instruments
4. DELETE_GROUPS members can delete instruments
5. Multiple group memberships are handled correctly

### Suggestions for Improvement

1. Create `src/casl/abilities/instruments.ability.spec.ts` with tests for:
   ```typescript
   describe('InstrumentAbility', () => {
     describe('Unauthenticated user', () => {
       it('can read instruments');
       it('cannot create instruments');
       it('cannot update instruments');
       it('cannot delete instruments');
     });

     describe('Authenticated user without special groups', () => {
       it('can read instruments');
       it('cannot create instruments');
       it('cannot update instruments');
       it('cannot delete instruments');
     });

     describe('User in ADMIN_GROUPS', () => {
       it('can read instruments');
       it('can create instruments');
       it('can update instruments');
       it('cannot delete instruments');
     });

     describe('User in DELETE_GROUPS', () => {
       it('can read instruments');
       it('cannot create instruments');
       it('cannot update instruments');
       it('can delete instruments');
     });

     describe('User in both ADMIN_GROUPS and DELETE_GROUPS', () => {
       it('can read, create, update, and delete instruments');
     });
   });
   ```

2. Add integration tests in `casl-ability.factory.spec.ts` to verify the `instrumentAccess` method delegates correctly to `InstrumentAbility`

---
## Security Examples

This section is empty as no relevant security vulnerabilities were identified.

---
## Testing for Security Use Cases

Not applicable as no security vulnerabilities were identified. However, for future reference:

### Testing Authorization Bypass Attempts
To test for potential authorization bypasses in the instruments authorization:

1. **Unauthenticated access to write operations**:
   - File: `src/casl/abilities/instruments.ability.ts:38-45`
   - Test: Verify that `buildAbility(null)` only allows `InstrumentRead`
   - Expected: `can(InstrumentCreate)`, `can(InstrumentUpdate)`, `can(InstrumentDelete)` should all return false

2. **Authenticated user without permissions**:
   - File: `src/casl/abilities/instruments.ability.ts:50`
   - Test: Verify that authenticated user with no special groups can only read
   - Expected: Only `InstrumentRead` should be allowed

3. **Group membership verification**:
   - File: `src/casl/abilities/instruments.ability.ts:52-67`
   - Test: Verify that group checks use `currentGroups.some()` with proper array inclusion
   - Expected: Users must be in the exact group specified in `accessGroups.admin` or `accessGroups.delete`

4. **Configuration injection**:
   - File: `src/casl/abilities/instruments.ability.ts:22-25`
   - Test: Verify that missing or malformed `accessGroups` configuration doesn't cause errors
   - Expected: Falls back to empty object, denying all write operations

---
## Summary

The `refactor-casl-factory-instruments` branch successfully extracts instrument authorization logic into a dedicated `InstrumentAbility` class, following the established pattern used for datasets. The refactoring maintains all existing functionality while improving code organization and maintainability.

**Key Findings**:
- Logic correctness: Maintained with no functional changes
- Edge cases: Properly handled for all user types
- Security: No vulnerabilities identified; authorization logic remains secure
- Test coverage: Inadequate - missing dedicated tests for the new `InstrumentAbility` class
- Documentation: Comprehensive new documentation added

**Recommendations**:
1. Add unit tests for `InstrumentAbility` to verify the complete permission matrix
2. Consider adding tests for the `instrumentAccess` wrapper in `CaslAbilityFactory`
3. Ensure the documentation is reviewed for consistency with existing authorization docs

**Verdict**: The changes are well-structured, necessary, and follow best practices. The primary gap is test coverage for the new ability class.

Generated by Mistral Vibe.
Co-Authored-By: Mistral Vibe <vibe@mistral.ai>

**Review by nitrosx — APPROVED** [review]
